### DISCUSSION TITLE
Ability to configure a folder (sub-path) for Azure Blob Storage and AWS S3 caches

### DISCUSSION THREAD

**Discussion by Piedone** [SixLabors/ImageSharp.Web#351] [Show and tell]
`AzureBlobStorageImageCache` and I believe also `AWSS3StorageCache` store all cache files in a single container/bucket. This prevents tenant-specific clean-up in a multi-tenant app (i.e. an app that hosts multiple websites), since we can't easily gather which cache file is related to which tenant. So, if you want to clear the cache just for a given tenant, or delete a tenant, including its cache files, then you can't do that without keeping track of ImageSharp's cache files on your own.

For `PhysicalFileSystemCache`, there's the `CacheFolder` option [that we utilize in Orchard Core](https://github.com/OrchardCMS/OrchardCore/blob/6a1aec3e360d67d189e8ed8b3109f5767ba046ab/src/OrchardCore.Modules/OrchardCore.Media/Startup.cs#L157C29-L157C40) to save cache files into tenant-specific folders. It would be great to have something similar for `AzureBlobStorageImageCache` and `AWSS3StorageCache` too.

I'm open to contribute this if the ImageSharp team wants the feature.

BTW added a feature request to Orchard Core about adding these features there too: https://github.com/OrchardCMS/OrchardCore/issues/15016.

**Comment by JimBobSquarePants** [comment]
I originally considered (in my usage of azure) the root container as the equivalent of a folder. Are there additional costs involved that I missed?

**Reply by Piedone** [reply]
Thanks for your reply! It's not really a cost thing, rather an organization/maintenance one.

With Orchard Core, we usually have a single storage account where we keep all the tenants' data. This currently entails a `dataprotection` (for key rings) and a `media` (for standard media files) container, perhaps also a `shells` one (for tenant settings). Ideally, we'd have an additional `imagesharpcache` one.

This way, it's easy to browse these containers on the Azure Portal, and easy to copy them from the production to the staging environment (and easy to back them up, though that wouldn't be necessary for the IS cache).

Having one container per tenant is possible (and I've also implemented it under https://github.com/OrchardCMS/OrchardCore/pull/15028), but would either add a lot of throwaway containers to such a storage account, or we'd need a separate storage account for just these cache containers. Due to the [rps limit set per storage account](https://learn.microsoft.com/en-us/azure/storage/common/scalability-targets-standard-account?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json#scale-targets-for-standard-storage-accounts:~:text=Default%20maximum%20request%20rate%20per%20storage%20account) the latter might be beneficial for apps with heavy usage, it looks like overkill for smaller apps.

**Reply by JimBobSquarePants** [reply]
OK. For the sake of flexibility and feature matching with the physical cache I'm game for the changes. Would you be able to submit a PR? My AWS experience is extremely limited.

**Reply by Piedone** [reply]
Great then! Yes, I'll work on a PR and hope to submit it next week.

**Reply by Piedone** [reply]
Opened a PR: https://github.com/SixLabors/ImageSharp.Web/pull/353 I have some questions, please see my notes and comments there.
